### ISSUE TITLE
Redundanter Code in der Umsatzverschlüsselung?

### ISSUE THREAD

**Issue by chenz-svsarrazin** [a-sit-plus/at-registrierkassen-mustercode#196]
Verstehe ich den Code (u.a. in `encryptCTR(...)`) falsch, oder ist das Einfügen des serialisierten Zählers in den Puffer redundant ausgeführt?

    //initialisation of the data which should be encrypted
    final ByteBuffer byteBufferData = ByteBuffer.allocate(16);
    byteBufferData.putLong(turnoverCounter);
    final byte[] data = byteBufferData.array();

    //now the turnover counter is represented in two's-complement representation (negative values are possible)
    //length is defined by the respective implementation (min. 5 bytes)
    byte[] turnOverCounterByteRep = get2ComplementRepForLong(turnoverCounter,turnOverCounterLengthInBytes);

    //two's-complement representation is copied to the data array, and inserted at index 0
    System.arraycopy(turnOverCounterByteRep,0,data,0,turnOverCounterByteRep.length);

Soweit ich das sehe, wird der Zähler zunächst mit `putLong()` eingefügt, und dann nochmal mit dem Ergebnis von `get2ComplementRepForLong()` (was mehr oder weniger das selbe macht) überschrieben?

**Comment by asitplus-pteufl** [comment]
Wenn der Umsatzzähler 8 Byte hat, dann ist der Code redundant, weil das putLong in Java eine 2erKomp. Darstellung in 8 byte durchführt. (also es ändert sich dann nichts mehr).
Wenn Sie aber einen Umsatzzähler von < 8 Byte haben, dann wird der mit der zweiten Methode korrekt abgebildet.

**Comment by chenz-svsarrazin** [comment]
Bei weniger als 8 Byte müsste aber dann Müll in `data` übrigbleiben, zb für counter = 1, length = 5

    // putLong(1)
    data = [0, 0, 0, 0, 0, 0, 0, 1, ...]

Bei length 5 würden danach nur die ersten 5 Bytes überschrieben, verschlüsselt würde dann

    // byteRep = [0, 0, 0, 0, 1]
    // arraycopy byteRep nach data, lenght 5
    data = [0, 0, 0, 0, 1, 0, 0, 1, ...]

Oder spielt das keine Rolle, weil dann letztendlich nur die ersten 5 Byte aus dem Ciphertext übernommen werden?

**Comment by asitplus-pteufl** [comment]
Genau, weil dann nur die 5 byte extrahiert werden (das geht aber nur im CTR Modus so).

Es wäre aber sicher schöner, wenn man das data nicht wiederverwenden würde. Ich lasse dieses Issue offen, damit das in einer weiteren Version schöner gemacht werden kann. Danke für den Hinweis!

**Comment by Alex1130** [comment]
* Warum ist dieser Schritt wichtig:
`````
System.arraycopy(turnOverCounterByteRep,0,data,0,turnOverCounterByteRep.length);
`````
könnte man auch
`````
final byte[] encryptedTurnOverValueComplete = cipher.doFinal(turnOverCounterByteRep);
`````
?
* Im Democode wird zuerst extrahiert und dann verschlüsselt, richtig?
In Detailfragen v1.2  2.5.5 ist das umgekehrt:
`````
Dabei wird wie folgt vorgegangen: Das Resultat des Verschlüsselungsprozesses (Prozess 2.5.3
oder Prozess 2.5.4) entspricht dem verschlüsselten Umsatzzähler (16-Bytes/128 Bits langer
Wert). Startend mit Byte 0 werden N Bytes aus dem Array extrahiert und BASE64-kodiert. N
entspricht dabei der im vorigen Prozess definierten Länge des verschlüsselten Umsatzzählers.
Der Minimalwert für N ist 5. Dies entspricht 5 Bytes/40 Bits, die aus dem verschlüsselten
Umsatzzähler extrahiert werden.
`````
